Level-Y/code.go

```diff
@@ -0,0 +1,91 @@
+///////////////////////////////////////////////////////////////////
+///                                                      		///
+///   0. code_test.go is passing but the code is vulnerable  	///
+///   1. Review the code. Can you spot the bugs(s)?       		///
+///   2. Fix the code.go but ensure that code_test.go passes    ///
+///   3. Run hack_test.go and if passing then CONGRATS!       	///
+///   4. If stuck then read the hint                     		///
+///   5. Compare your solution with solution/solution.go   		///
+///                                                      		///
+///////////////////////////////////////////////////////////////////
+
+package main
+
+import (
+	"encoding/json"
+	"log"
+	"net/http"
+	"regexp"
+)
+
+// Test users
+var testFakeMockUsers = map[string]string{
+	"user1@example.com": "password12345",
+	"user2@example.com": "B7rx9OkWVdx13$QF6Imq",
+	"user3@example.com": "hoxnNT4g&ER0&9Nz0pLO",
+	"user4@example.com": "Log4Fun",
+}
+
+var reqBody struct {
+	Email    string `json:"email"`
+	Password string `json:"password"`
+}
+
+func isValidEmail(email string) bool {
+	// The provided regular expression pattern for email validation by OWASP
+	// https://owasp.org/www-community/OWASP_Validation_Regex_Repository
+	emailPattern := `^[a-zA-Z0-9_+&*-]+(?:\.[a-zA-Z0-9_+&*-]+)*@(?:[a-zA-Z0-9-]+\.)+[a-zA-Z]{2,}$`
+	match, err := regexp.MatchString(emailPattern, email)
+	if err != nil {
+		return false
+	}
+	return match
+}
+
+func loginHandler(w http.ResponseWriter, r *http.Request) {
+	if r.Method == "POST" {
+
+		decode := json.NewDecoder(r.Body)
+		decode.DisallowUnknownFields()
+
+		err := decode.Decode(&reqBody)
+		if err != nil {
+			http.Error(w, "Cannot decode body", http.StatusBadRequest)
+			return
+		}
+		email := reqBody.Email
+		password := reqBody.Password
+
+		if !isValidEmail(email) {
+			log.Printf("Invalid email format: %q", email)
+			http.Error(w, "Invalid email format", http.StatusBadRequest)
+			return
+		}
+
+		storedPassword, ok := testFakeMockUsers[email]
+		if !ok {
+			http.Error(w, "invalid email or password", http.StatusUnauthorized)
```

Comment:
I wonder if it's obvious enough that this error is different to the password one (Just the capital letters change).
I would change it to:
```suggestion
			http.Error(w, "Invalid email", http.StatusUnauthorized)
```
That way the user enumeration vuln is more clear (I'm not sure how difficult this level is supposed to be tho)

Reply:
The error message was based on a vulnerability that I found and that was fixed. 
I thought it would be cool to use a real case. If it's not obvious enough I can commit the suggestion. 
Would love to have the point of view of @jkcso on that too.